C/services/south/ingest.cpp

```diff
@@ -28,6 +28,157 @@ static void ingestThread(Ingest *ingest)
 	}
 }
 
+/**
+ * Create a row for given assetName in statistics DB table, if not present already
+ * The key checked/created in the table is "SOUTH_READINGS_FROM_<assetName>"
+ * 
+ * @param assetName     Asset name for the plugin that is sending readings
+ */
+int Ingest::CreateStatsDbEntry(const string& assetName)
+{
+	// Prepare foglamp.statistics update
+	string statistics_key = "south_readings_from_" + assetName;
+	for (auto & c: statistics_key) c = toupper(c);
+	
+	// SELECT * FROM foglamp.configuration WHERE key = categoryName
+	const Condition conditionKey(Equals);
+	Where *wKey = new Where("key", conditionKey, statistics_key);
+	Query qKey(wKey);
+
+	ResultSet* result = 0;
+	try
+	{
+		// Query via storage client
+		result = m_storage.queryTable("statistics", qKey);
+
+		if (!result->rowCount())
+		{
+			// Prepare insert values for insertTable
+			InsertValues newStatsEntry;
+			newStatsEntry.push_back(InsertValue("key", statistics_key));
+			newStatsEntry.push_back(InsertValue("description", string("Readings received from asset ")+assetName));
+			// Set "value" field for insert using the JSON document object
+			newStatsEntry.push_back(InsertValue("value", 0));
+			newStatsEntry.push_back(InsertValue("previous_value", 0));
+
+			// Do the insert
+			if (!m_storage.insertTable("statistics", newStatsEntry))
+			{
+				m_logger->error("%s:%d : Insert new row into statistics table failed, newStatsEntry='%s'", __FUNCTION__, __LINE__, newStatsEntry.toJSON().c_str());
+				return -1;
+			}
+			else
+				m_logger->info("%s:%d : Inserted new row into statistics table, newStatsEntry='%s'", __FUNCTION__, __LINE__, newStatsEntry.toJSON().c_str());
+		}
+	}
+	catch (...)
+	{
+		m_logger->error("%s:%d : Unable to create new row in statistics table with key='%s'", __FUNCTION__, __LINE__, statistics_key.c_str());
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * Thread to update statistics table in DB
+ */
+static void statsThread(Ingest *ingest)
+{
+	while (ingest->running())
```

Comment:
The thread should sleep for a while, on the Python side there is a configuration entry for how long between statistics updates.

Reply:
ingest->updateStats() basically waits on a condition variable - m_statsCv. And this condition variable is notified by the ingest/other thread at the end of processQueue function. So this thread also sleeps for around same time as the ingest thread. And that time comes from config->maxSendLatency. That's 5 seconds by default. I initially implemented a separate sleeping time for this thread, but then, any snapshot of statistics table won't reflect a consistent picture (input from Praveen). So implemented the way it is presently, same sleep time for both threads.

Reply:
OK, thanks - missed that